LdapForNet/LdapEntry.cs

```diff
@@ -1,287 +1,352 @@
+
+        public int GetUserPrimaryID()
+        {
+            var objectSid = GetAttribute(LdapAttributes.ObjectSid)?.GetValue<byte[]>();
+            if (objectSid != null)
+                return BitConverter.ToInt32(objectSid, objectSid.Length - 4); //last 4 bytes are primary group id
+
+            return -1;
+        }
+
+        public string GetString(string attributeName) => GetAttribute(attributeName)?.GetValue<string>();
+
+        public byte[] GetBytes(string attributeName) => GetAttribute(attributeName)?.GetValue<byte[]>();
+
+        public IEnumerable<string> GetStrings(string attributeName) => GetAttribute(attributeName)?.GetValues<string>() ?? Enumerable.Empty<string>();
+
+        public IEnumerable<byte[]> GetByteArrays(string attributeName) => GetAttribute(attributeName)?.GetValues<byte[]>() ?? Enumerable.Empty<byte[]>();
+    }
+
+    public class LdapModifyEntry
+    {
+        public string Dn { get; set; }
+        public List<LdapModifyAttribute> Attributes { get; set; }
+    }
+
+    public class LdapModifyAttribute
+    {
+        public string Type { get; set; }
+        public List<string> Values { get; set; }
+
+        public Native.Native.LdapModOperation LdapModOperation { get; set; } =
+            Native.Native.LdapModOperation.LDAP_MOD_REPLACE;
+    }
+
+    public class DirectoryAttribute
+    {
+        private List<byte[]> _byteValues;
+        private List<string> _stringValues;
+
+        public string Name { get; set; }
+
+        public T GetValue<T>()
+            where T : class, IEnumerable
+        {
+            var items = GetValues<T>();
+            var item = items.FirstOrDefault();
+            return item == default(T) ? default : item;
+        }
+
+        public IReadOnlyList<T> GetValues<T>() where T : class, IEnumerable
+        {
+            var type = typeof(T);
+
+            if (type == typeof(string))
+            {
+                if (_stringValues == null)
+                {
+                    if (_byteValues == null)
+                    {
+                        _stringValues = new List<string>(0);
+                    }
+                    else
+                    {
+                        _stringValues = new List<string>(_byteValues.Count);
+                        for (int i = 0; i < _byteValues.Count; i++)
+                        {
+                            _stringValues.Add(Encoder.Instance.GetString(_byteValues[i]));
+                        }
+                    }
+                }
+
+                return (IReadOnlyList<T>)_stringValues;
+            }
+
+            if (type == typeof(string))
+            {
+                if (_byteValues == null)
+                {
+                    if (_stringValues == null)
+                    {
+                        _byteValues = new List<byte[]>(0);
+                    }
+                    else
+                    {
+                        _byteValues = new List<byte[]>(_stringValues.Count);
```

Comment:
please use more elegant way to initialize values
```

Reply:
It is maybe more elegant, but it is also very costly. Each Select will need to instantiate an Enumerable. Then, the ToList will instantiate a List and resize it for each element of the Enumerable (since it can't know the wanted size at the instantiation).
In the code commited, I defined the size at the instanciation so the memory allocation for the List is only made once for a GetValues call, which is faster